### ISSUE TITLE
Use `Object.keys` / `Object.values` to succinctly appease Typescript

### ISSUE THREAD

**Issue by scheibo** [Zarel/Pokemon-Showdown#5281]
@KrisXV raised this idea in chat:

```javascript
for (let i in template.abilities) {
  // @ts-ignore
  if (set.ability === template.abilities[i]) legalAbility = true;
}
```

could instead be:

```javascript
for (let abilityName of Object.values(template.abilities)) {
  if (set.ability === abilityName) legalAbility = true;
}
```

We already use `Object.keys`/`Object.values` in the codebase, so needing to polyfill shouldn't be an issue. The alternative pattern of using `/** @type {...} */` or declaring the loop variable outside of the loop to give it a type is more verbose.

Performance is somewhat of a concern, but overall, I think we should value readability and succintness and only optimize the looping pattern if necessary in the tightest loops around hotspots (using real benchmarks from PS! instead of microbenchmarks).

**Comment by scheibo** [comment]
![Data](https://media.nesta.org.uk/images/in-god-we-trust.max-1200x600.jpg)

https://jsperf.com/ps-object-values-vs-for-in

Seems to be really poor :(

**Comment by Slayer95** [comment]
@scheibo, your benchmark says ``in Object.values()``, while it ought to say ``of Object.values()``

**Comment by Zarel** [comment]
```js
for (let abilityName of Object.values(template.abilities)) {
  if (set.ability === abilityName) legalAbility = true;
}
```

can be further simplified to

```js
const legalAbility = Object.values(template.abilities).some(
  ability => ability === abilityName
);
```

**Comment by scheibo** [comment]
Good catch @Slayer95 . Thats makes things more fair, but still quite a bit slower.

**Comment by Slayer95** [comment]
Better yet: ``const legalAbility = Object.values(template.abilities).includes(abilityName);`` :)

**Comment by scheibo** [comment]
As fun a code golf is, I think the important thing to decide is whether we'd support switching many of our `for...in` loops to use `Object.values` for increased succinctness and readability. Basically, are we willing to incur a small latency hit in exchange for nicer code? I'm leaning yes (because we can always switch back in places where it matters *if* it matters). If @Zarel signs off I think @KrisXV is willing to start work on this.

**Comment by Slayer95** [comment]
I support switching to ``Object.values()`` for the better readability from their implied variable names (``abilityName`` is better than ``abilities[i]``).

But I wonder, why is this even an issue for TypeScript? Does TS not know that ``for..in`` variables are always of type ``String``? Or does it not know the types of the values? ?___?

**Comment by Slayer95** [comment]
The bigger concern, however, is that ``Object.(keys|values)`` should increase the GC strain, which is a major concern for PS. Do we have a way to measure its impact there?

**Comment by Zarel** [comment]
Wow I completely forgot about the existence of `includes`.

**Comment by Zarel** [comment]
I'd say "yes outside of inner loops".

**Comment by Zarel** [comment]
GC strain is only a problem in the main process; sim processes should be fine.

**Comment by Zarel** [comment]
@Slayer95 The problem is that TypeScript doesn't support closed types:

https://github.com/Microsoft/TypeScript/issues/3500#issuecomment-388883320

In other words, TypeScript considers `{atk: 1, hasAttack: true}` to satisfy the type `{atk: number}`, so iterating `{atk: number}` is not a guarantee that `atk` is the only key. It's impossible to use an object like a dictionary unless you drop all safety and define the type as `{[k: string]: number}`.

**Comment by Zarel** [comment]
TypeScript also intentionally makes it _very_ hard to get around this with type assertions, for reasons that are incredibly confusing to me.

**Comment by Zarel** [comment]
One possible workaround:

```ts
type StatsTable = {
	hp: number, atk: number, def: number, spa: number, spd: number, spe: number,
	[k: string]: number | undefined,
};
```

This allows the assertion to be a lot easier (just `stats[stat]!` in a `for...in` loop), in exchange for certain bugs (like `stats.attk === 100` or ``print(`atk: ${stats.attk}`)``) not being detectable. _Most_ accesses that aren't an equality check or string concatenation would still be detectable as a bug, though.

Thoughts, @scheibo?

**Comment by scheibo** [comment]
Not a fan. Unless I'm mistaken, you're throwing away the majority of the type safety for convenience here, IMO. Which accesses exactly would be detectable as a bug? The ones which have a `!` are all suspect, sure, but its easy to miss one character and i imagine we'll be sprinkling it all through the codebase anyway, so its hardly going to jump out as atypical.

**Comment by scheibo** [comment]
One pattern that might be simpler and faster:

```javascript
// In globals or on Battle like NOT_FAIL etc
const STATS: Stats[] = ['hp', 'atk', 'def', 'spa', 'spd', 'spe'];

// ...

for (const stat of STATS) {
  pokemon.stats[stat] = // ...
}
```

this would probably cut down on a ton of the `// @ts-ignore`

**Comment by Zarel** [comment]
The client has `Dex.statNames` (`this.statNames` in a battle context). We could do the same thing on the server.

**Comment by Zarel** [comment]
`Dex.STATS` could also work. No strong preference.

**Comment by Zarel** [comment]
> Which accesses exactly would be detectable as a bug?

Any function call expecting a number, any comparison other than `===` (for instance, `>`), any math done on it such as multiplication... I actually think it covers most accesses.

**Comment by KrisXV** [comment]
I’ve been working on this btw, but I’ve had a power outage so it’s kind of hard

**Comment by scheibo** [comment]
> Any function call expecting a number, any comparison other than `===` (for instance, `>`), any math done on it such as multiplication... I actually think it covers most accesses.

Developers are going to try to access `this.stats[stat]` where `stat` is a string. And youre right, in these circumstances they'll get told off by the compiler. Only they'll "know" that `stat` *should* be defined, should they'll chuck on a `!` to shut up the "stupid compiler", only to be incorrect sometimes, and the pattern will be so prevalent it will be hard to tell where (note: in this example, *I* am the developer sometimes who thinks he is better at static analysis than the compiler).

`!` is simultaneously great because of how convenient and ubiquitous it is, but this is also its shortcoming. It normalizes 'ignoring the compiler' - which is sometimes perfectly valid and the right thing to do, but I dont think we should adopt a pattern in the codebase where it's the norm.

Kris's suggestion to use `Object.values`, my suggestion to use a `STATS` constant (or `Dex.STATS` like the client), or even the pattern ive been using where I declare to loop variable to be a `StatName` outside of the loop all are correct and safe, so I dont know why we would pursue any less safe path when its not like we dont have options.

**Comment by Slayer95** [comment]
So is this approach accepted? @Zarel

**Comment by scheibo** [comment]
c075cad379c1b85dcd2ad16f36526ea5d4dc3377 added it to TSLint this for Typescript, because in that case we can use it to avoid `for-in` for arrays only. Do want to disable all `for-in` usage?

**Comment by Slayer95** [comment]
Yes, that's what I am proposing, since TypeScript doesn't play well with ``for-in`` (it seems to me that iteration *a la* JS ``for-in`` is an unusual pattern in most languages?)

**Comment by scheibo** [comment]
I don't think a global ban on the construct in possible with TSLint? I think we might need to use https://www.npmjs.com/package/tslint-eslint-rules to cover that?

**Comment by Slayer95** [comment]
I think it would be better to use [ESLint to lint our TS files](https://eslint.org/blog/2019/01/future-typescript-eslint) as well, given that TSLint is on [road to deprecation](https://github.com/Microsoft/TypeScript/issues/29288)

**Comment by Zarel** [comment]
@Slayer95, I saw that announcement, too, but I couldn't see how to use it at all. As far as I can tell, it's just a project on their roadmap?

**Comment by Zarel** [comment]
@Slayer95 

> So is this approach accepted?

The `Dex.STATS` constant sounds good. I'm also fine with `Object.values` for various stats/avatar tables.

I might want to keep `for...in` around for other things, though. Some things _are_ in inner loops!

> Going one step further, we would add `"ForInStatement"` to ESLint's `no-restricted-syntax` feature.

No. :[

> Yes, that's what I am proposing, since TypeScript doesn't play well with `for-in` (it seems to me that iteration _a la_ JS `for-in` is an unusual pattern in most languages?)

Iterating dictionaries is definitely not an unusual pattern in most languages. :[ Our main problem is that we use JavaScript Objects for certain dictionary-like data structures, namely stats tables and ability tables.

I might be wrong, but I think those are the only ones?

**Comment by Zarel** [comment]
After further googling, it appears `typescript-eslint` is usable. Feel free to set it up!

**Comment by Slayer95** [comment]
- ``Dex.stats.ids()`` is now available
- ``typescript-eslint`` is now in use.
